ssx: Checkpoint mutex

Lazin opened the pull request:
This PR implements a `checkpoint_mutex`. The mutex can remember the last callsite that was able to acquire it. This information can be accessed in order to generate the diagnostic dump. It is also included into exception messages.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin committed:
ssx: Add mutex with checkpointing mechanism

The mutex is named and works pretty much the same way our mutex works.
The difference is that every time the mutex is acquired the caller has
to pass a string that contains the name of the callsite.

If the mutex is acquired the checkpoint is created. The checkpoint is
simply a name of the callsite which was able to acquire the mutex and
the timestamp. This information is stored in the mutex until the mutex
is releasaed. When this happens the checkpoint is erased.

If any error occured the checkpoint information is included into the
exception.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
archival: Use checkpoint_mutex in the ntp_archiver

Change _mutex and _uploads_active semaphores to be a cp-mutex. Update
acquire callsites.

The goal here is to have more visibility in a situation when something
got stuck.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

mmaslankaprv commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,307 @@
+
+    bool has_units() const noexcept { return _sem.available_units() > 0; }
+
+    basic_mutex_checkpoint<Clock> get_blocking_checkpoint() const noexcept {
```
i am wondering if that should return optional ?

Lazin replied:
done

mmaslankaprv commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,307 @@
+/// was awaited. In the event of a timeout the exception will contain the
+/// name of the checkpoint that holds the units.
+template<typename Clock = seastar::timer<>::clock>
+class basic_checkpoint_mutex {
```
It would be great if this `mutex` had the same api as our current mutex for it be drop in replacement, here the API is completely different.

Lazin replied:
done

mmaslankaprv commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,307 @@
+seastar::future<basic_mutex_units<Clock>> get_units_with_checkpoint(
+  basic_checkpoint_mutex<Clock>& mut,
+  typename Clock::time_point deadline,
+  seastar::sstring checkpoint_name) {
```
can we have a set of methods that would use source file location as a checkpoint name ?

mmaslankaprv replied:
Having the same API as current mutex and using location as the name would allow us to replace some critical mutexes in the code with this implementation

rockwotj replied:
I agree

Lazin replied:
updated the pr
the interface now matches the `utils/mutex`
the methods are using `vlog::file_line` instead of string

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--243633a0-7c9c-4621-b094-ee2e0e79dad3-->
<!--TEST RESULTS-->
<!--f6a71252-e619-4223-9bd1-85758372033c-->
<!--TEST RESULTS-->
<!--1f9fa0ff-27a5-4985-96bc-a66b5eeca9a7-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67166">build#67166</a></summary>


| test_class               | test_method                  | test_arguments                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-----------------------------|:-----------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| FollowerFetchingTest     | test_basic_follower_fetching | {"read_from_object_store": true}                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67166#01975ebf-ae71-4882-a260-4bff52c84969 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations         | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67166#01975ea8-7f3e-4adb-9ef4-a2216f419410 | FLAKY         | 19/21    | upstream reliability is '99.45054945054946'. current run reliability is '90.47619047619048'. drift is 8.97436 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67210">build#67210</a></summary>


| test_class                  | test_method                | test_arguments                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:----------------------------|:---------------------------|:----------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest       | test_rack_awareness        |                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67210#0197608a-57eb-41a0-ac73-84d2d7d9aa24 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| SaslPlainTest               | test_plain_authn           | {"client_type": 3, "sasl_plain_enabled": true, "scram_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67210#0197608a-57ec-4e3c-895d-2f879f6579c4 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2}                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67210#019760a8-0d50-4071-ac05-802ff2abdd8b | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67294">build#67294</a></summary>


| test_class                  | test_method                                             | test_arguments                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:--------------------------------------------------------|:---------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DeleteRecordsTest           | test_delete_records_empty_or_missing_topic_or_partition | {"cloud_storage_enabled": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67294#01976713-e571-49eb-b65a-a99f9764dd04 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| FullDiskTest                | test_full_disk_no_produce                               |                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67294#01976713-e571-49eb-b65a-a99f9764dd04 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| OffsetForLeaderEpochTest    | test_offset_for_leader_epoch_transfer                   |                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67294#0197669c-e79e-4afb-9541-51895d5af3de | FLAKY         | 20/21    | upstream reliability is '98.23008849557522'. current run reliability is '95.23809523809523'. drift is 2.99199 and the allowed drift is set to 50. The test should PASS |
| PartitionReassignmentsTest  | test_add_partitions_with_inprogress_reassignments       |                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67294#0197669c-e7a1-474f-bc37-ee39ca82c073 | FLAKY         | 20/21    | upstream reliability is '93.16109422492401'. current run reliability is '95.23809523809523'. drift is -2.077 and the allowed drift is set to 50. The test should PASS  |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test                              | {"cloud_storage_type": 2}        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67294#01976713-e570-49ac-a299-262e1094193a | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
</details>
<!--/body-->

dotnwat commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,380 @@
+    }
+
+    void release() noexcept {
+        _sem.signal(1);
```
nit: might be a nice place to add another assert that the mutex is held?

Lazin replied:
done

dotnwat commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,380 @@
+      : _mut(std::exchange(other._mut, nullptr)) {
+        // Move constructor transfers ownership of the mutex.
+        // The moved-from object will not hold any mutex.
+        if (_mut) {
+            vassert(!_mut->has_units(), "The mutex is not acquired");
+        }
```
curious if this check catches something specific. for example, presumably the moved-from object was valid to begin with..

Lazin replied:
It just checks the invariant of the 'units' object. If the `_mut` is not `nullptr` then the mutex couldn't have the units because the units are acquired and the 'units' object represents them and will try to release them in d-tor.

dotnwat commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,380 @@
+auto basic_checkpoint_mutex<Clock>::with(
+  Func&& func, vlog::file_line line) noexcept {
+    auto units = co_await get_units(line);
+    std::invoke(std::forward<Func>(func));
```
shouldn't these `return std::invoke` ... ? i also don't see any coverage for these in the tests. did i miss them?

Lazin replied:
that makes sense, I'll update the code

dotnwat commented on src/v/ssx/tests/checkpoint_mutex_test.cc:
```diff
@@ -0,0 +1,193 @@
+
+    // Sleep long enough to ensure that the timeout occurs.
+    ss::sleep(100ms).get();
+    BOOST_REQUIRE(mutex.has_units() == false);
```
let's write new tests with google test

Lazin replied:
switched to google test

dotnwat commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,380 @@
+                  "{}: {} at {}",
+                  name,
+                  _current.value().line,
+                  _current.value().time.time_since_epoch().count());
```
I don't think this works as expected with all clocks. For example the default clock, `seastar::timer<>::clock` is `chrono::steady_clock` and its epoch is not necessarily the Unix epoch. In particular, I dont' think there is really any relationship between the timer the mutex uses, and the desire to track a time point. Presumably we can just always use `lowres_system_clock` for this?

Lazin replied:
I updated the formatting. IIRC the step is 10ns or 100ns with the lowres clock. But it's still pretty easy to infer what is the right timestamp there.

dotnwat replied:
The issue wasn't the formatting, but forcing milliseconds is great because I ran this yesterday, and I'm pretty sure I was getting either microseconds or nanoseconds.

There are two issues I was trying to convey here:

1. the usefuless of `time_since_epoch` in the exception (assuming Unix epoch)
2. the impact of steady_clock on the value of `time_since_epoch`

I'm assuming the information we want to deduce from the exception message is how long the lock has been held. If we print out `time_point_of_acquire.time_since_epoch()` then is this sufficient? It doesn't seem like it: we'd need another reference point, for example the timestamp of the log message (or if we were examining the state in a core dump, i'd guess we'd need the exact time of the core dump?). Should we directly report the hold duration (and potentially the timepoint as well)?

The second issue I wanted to raise was the use of `time_since_epoch` in conjunction with a steady_clock (the default clock of seastar timer). The `time_since_epoch` interface doesn't return the time since Unix epoch, it returns the time since the clock's zero point, and for steady_clock, the zero point is undefined. In practice it is the time since the kernel booted, but in principle it could be a random value. That makes it even more difficult to work with because we can't compare it to a log message timestamp.

So my hunch here is that we just want to use `lowres_system_clock` or whatever cheap source of wall clock time exists?

Am I way off base here, or is the steady clock and time_since_epoch sufficient for what we are trying to acheive?

Lazin replied:
fixed that
I'm adding number of ms since the mutex was acquired

dotnwat commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,380 @@
+            return checkpoint_mutex_broken("Broken mutex, " + _name, _cs);
+        }
+        checkpoint_mutex_aborted aborted() const noexcept {
+            return checkpoint_mutex_aborted("Mutex aborted, " + _name, _cs);
```
nit: `Mutex wait aborted`

Lazin replied:
done

dotnwat reviewed (commented):
good stuff !

dotnwat commented on src/v/ssx/tests/checkpoint_mutex_test.cc:
```diff
@@ -0,0 +1,189 @@
+    EXPECT_EQ(
+      mutex.get_blocking_checkpoint()->line.filename,
+      ss::sstring("checkpoint_mutex_test.cc"));
+    EXPECT_EQ(mutex.get_blocking_checkpoint()->line.line, 32);
```
i was messing around with the test, and any change to headers makes all these line number assertions fail. you can fix them like this

```
diff --git a/src/v/ssx/tests/checkpoint_mutex_test.cc b/src/v/ssx/tests/checkpoint_mutex_test.cc
index 89982fa17b..8794f65862 100644
--- a/src/v/ssx/tests/checkpoint_mutex_test.cc
+++ b/src/v/ssx/tests/checkpoint_mutex_test.cc
@@ -29,13 +29,15 @@ TEST(CheckpointMutex, checkpoint_mutex_lock) {
     EXPECT_TRUE(mutex.get_blocking_checkpoint() == std::nullopt);

     // Acquire the mutex normally.
+    auto line = vlog::file_line::current();
     auto units = mutex.get_units().get();
     EXPECT_TRUE(mutex.has_units() == false);
     EXPECT_EQ(
       mutex.get_blocking_checkpoint()->line.filename,
       ss::sstring("checkpoint_mutex_test.cc"));
-    EXPECT_EQ(mutex.get_blocking_checkpoint()->line.line, 32);
+    EXPECT_EQ(mutex.get_blocking_checkpoint()->line.line, line.line+1);

+    line = vlog::file_line::current();
     auto fut = mutex.get_units();
     EXPECT_TRUE(fut.available() == false);

@@ -48,7 +50,7 @@ TEST(CheckpointMutex, checkpoint_mutex_lock) {
     EXPECT_EQ(
       mutex.get_blocking_checkpoint()->line.filename,
       ss::sstring("checkpoint_mutex_test.cc"));
-    EXPECT_EQ(mutex.get_blocking_checkpoint()->line.line, 39);
+    EXPECT_EQ(mutex.get_blocking_checkpoint()->line.line, line.line+1);

     units2.release();
     EXPECT_TRUE(mutex.has_units() == true);
```

Lazin replied:
done

dotnwat commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,385 @@
+
+template<typename Clock>
+template<typename Func>
+auto basic_checkpoint_mutex<Clock>::with(
```
afaict there isn't any code using these `with` interfaces (add tests?), but i don't think this will even compile if there were users, because afaict coroutines cannot have placeholder return types like `auto` (you'd need to add `-> seastar::future<>` annotation, but then it's not a placeholder type).

Lazin replied:
fixed

dotnwat commented on src/v/ssx/checkpoint_mutex.h:
```diff
@@ -0,0 +1,385 @@
+auto basic_checkpoint_mutex<Clock>::with(
+  Clock::duration timeout, Func&& func, vlog::file_line line) noexcept {
+    auto units = co_await get_units(timeout, line);
+    co_return std::invoke(std::forward<Func>(func));
```
this also won't compile in the vast majority of cases, where `func` returns a future (it will if func returns a non-future value, but using a mutex around non-blocking code is rare).

Lazin replied:
I'll check, this doesn't have test btw

Lazin replied:
I think I fixed it. It works in test.

dotnwat reviewed (commented):
add a few new review comments, but also expanded my note about time_since_epoch.

dotnwat reviewed (commented):
it all looks good, but i think you want to replace `std::invoke` with `seastar::futurize_invoke` to handle whatever kind of `Func` we throw at `with(..)`

dotnwat committed:
ssx: make tests fail with early units release

The tests were not strong enough because the function that tested the
mutex state executed within the lambda containing the held units.
Howevever in real world code the function will return a future that may
execute long after the units were released.

This modified test will fail without the fix in the previous commit.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

Lazin commented:
> it all looks good, but i think you want to replace `std::invoke` with `seastar::futurize_invoke` to handle whatever kind of `Func` we throw at `with(..)`

fixed that 
cc @dotnwat

dotnwat committed:
ssx: do not early release units

Prior to this commit the units were not held for the duraiton of the
critical seciton.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ssx: use gtest richer primitives

Provides better reporting and clearer meaning.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
ssx: apply clang format

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat reviewed (commented):
still was a bug in the `with(...)` implementation. i pushed some patches to this PR that (1) fix the bug and (2) adds a test to demonstrate the bug.

lemme know what you think.

dotnwat commented on src/v/cluster/archival/ntp_archiver_service.cc:
```diff
@@ -3610,11 +3611,11 @@ ntp_archiver::prepare_transfer_leadership(ss::lowres_clock::duration timeout) {
     ss::gate::holder holder(_gate);
     try {
-        co_await ss::get_units(_uploads_active, 1, timeout);
+        auto units = co_await _uploads_active.get_units(timeout);
         vlog(
           _rtclog.trace,
           "prepare_transfer_leadership: got units (current {})",
-          _uploads_active.current());
+          _uploads_active.has_units());
```
it looks here like the semantics of the print statement will be differnet. before, `.current` would not reflect the acquisition above because it was immediately released. now it does, because you're keeping an raii object alive.

Lazin replied:
it doesn't matter, I don't think we need to log units here at all
